Gems/AtomTressFX/Code/Passes/HairSkinningComputePass.cpp

```diff
@@ -0,0 +1,277 @@
+                {
+                    return;
+                }
+
+
+                // For dispatches that change the same base data this lock method might not be enough.
+                // Since build might take more than several frames, iterating over passes in the future
+                // is recommended to verify that a proper barrier is put in place.
+                // Currently this approach of blocking set of passes is not taken anywhere in Atom and
+                // we leave it as low importance as it is only relevant rarely when hot reloading changed
+                // compute shaders that depends on a resource from previous pass that are also compiled.
+                // To that affect, iterating over the entire simulation pass chain rather than iterating
+                // over each pass separately is safer for simulation integrity after shader changes.
+                uint32_t iterations = m_allowSimIterations ? AZ::GetMax(hairObject->GetCPULocalShapeIterations(), 1) : 1;
+                AZStd::lock_guard<AZStd::mutex> lock(m_mutex);
+
+                for (uint32_t j = 0; j < iterations; ++j)
+                {
+                    m_dispatchItems.insert(dispatchItem);
+                }
+            }
+
+            void HairSkinningComputePass::BuildCommandListInternal(const RHI::FrameGraphExecuteContext& context)
+            {
+                if (m_buildShaderAndData)
+                {   // Protect against shader and data async change that were not carried out
+                    m_dispatchItems.clear();
+                    return;
+                }
+
+                RHI::CommandList* commandList = context.GetCommandList();
+
+                // The following will bind all registered Srgs set in m_shaderResourceGroupsToBind
+                // and sends them to the command list ahead of the dispatch.
+                // This includes the PerView, PerScene and PerPass srgs (what about per draw?)
+                SetSrgsForDispatch(commandList);
+
+                for (const RHI::DispatchItem* dispatchItem : m_dispatchItems)
+                {
+                    commandList->Submit(*dispatchItem);
+                }
+
+                // Clear the dispatch items. They will need to be re-populated next frame
+                m_dispatchItems.clear();
+            }
+
+            void HairSkinningComputePass::BuildShaderAndRenderData()
+            {
+                AZStd::lock_guard<AZStd::mutex> lock(m_mutex);
+                if (!AcquireFeatureProcessor())
+                {
+                    m_buildShaderAndData = true;
+                    return;
+                }
+
+                // Flag the FP not to add any more dispatches until shader rebuild was done
+                m_featureProcessor->SetAddDispatchEnable(false);
+
+                while (true)
+                {   // This loop bounds to terminate once the dispatches are submitted.
+                    if (m_dispatchItems.empty())
```

Comment:
I assume this is always being called in a job thread, and another thread is going to submit the dispatch items, which is why m_dispatchItems is eventually empty, is that correct? 

This looks like a recipe for a deadlock, since you also acquire a mutex above. If any other thread tries to acquire that mutex, and if that other thread prevents m_dispatchItems from being cleared, then it will get stuck here.
E.g. ThreadA is adding dispatch items. ThreadB calls BuildShaderAndRenderData. DispatchItems don't get cleared until ThreadA finishes, but ThreadA is waiting on ThreadB.

Reply:
yes and no.
Your point is valid and what you describe can indeed be a potential for a deadlock.
What I do and the reasons are actually different.
This case is only triggered when a shader required compilation (hot reload) and indeed I assume this is done in Async manner as you assumed.
In the case of hot reload request, I try to do something that no other system does so far - I'm invoking a full chain clean of all dispatches in the hair pass system, hence all compute shaders. 
I'm also being over protective by removing any existing until next frame, and sending for shader build (combined with the FP dispatch build), and verify that the FP itself is not in the middle of a switch (was not acquired).
 I'm doing all of this under the assumption that changing the shader will affect all the hair compute passes and not only the one affected.  This is NOT a closed problem but it's more than anything done so far in Atom. 

To your remark about the mutex - it is used to determine if to allow more dispatch additions.  Two methods that can be on different threads use it and both are short and do not depend on each other with a very decisive clear method taking place always at least once per frame.

Since this is a very valid concern, please check my logic and verify that indeed it solves the concern so we can remove this flag.

Reply:
My concern is that there is nothing in HairSkinningComputePass::BuildShaderAndRenderData that actually clears m_dispatchItems. It's creating an infinite loop, and waiting for something else to clear m_dispatchItems. But that may never happen, due to the mutex.

Let's say ThreadA is executing HairFeatureProcessor::Render, and there are 10 hair objects. It loops over each object, and calls HairSkinningComputePass::AddDispatchItem, once for each item, acquiring and releasing HairSkinningComputePass::m_mutex each time. Let's assume it gets through 5 items, so m_dispatchItems is not empty, but it's also not finished adding items, when an OnShader*Reinitialzed event occurs asynchronously in ThreadB. Now HairSkinningComputePass::BuildShaderAndRenderData is called, and it acquires a lock on HairSkinningComputePass::m_mutex in ThreadB. ThreadA cannot acquire the mutex, so it can't finish adding the remaining 5 hair objects. It has to wait for ThreadB to release, but ThreadB won't release the mutex until m_dispatchItems is empty, which won't happen until HairSkinningComputePass::BuildCommandListInternal is called, which won't ever happen because BuildCommandListInternal occurs after Render, and Render is stuck in ThreadA.

I think you can work around it in a few ways, but the best is probably to give the responsibility to the FeatureProcessor to wait for the right point in time to tell the compute pass to Reload, as @antonmic suggests below.

Reply:
you are correct - since the dispatches are being added one after another this is possible and will need a solution at the level above. tnx for catching this.